sql: implement CREATE TABLE ... AS ... (new feature).

WillHaack opened the pull request:
`CREATE TABLE <tableName> AS` allows a user to instantiate a table based off the results of a select statement. For example a user may write the command `CREATE TABLE sales AS SELECT item, stock, price FROM inventory WHERE stock > 0 AND price < 10` to create a table sales which has the columns item, stock, and price and the rows of all the rows in inventory matching the query.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8802)

<!-- Reviewable:end -->

WillHaack committed:
Added the CREATE TABLE ... AS feature to the grammar.

CREATE TABLE <tableName> AS allows a user to instantiate a table
based off the results of a select statement. For example a user may
write the command
CREATE TABLE sales AS SELECT item, stock, price FROM inventory WHERE
stock > 0 AND price < 10
to create a table sales which has the columns item, stock, and price
and the rows of all the rows in inventory matching the query.

WillHaack commented on sql/create.go:
```diff
@@ -440,18 +479,69 @@ func (n *createTableNode) Start() error {
+		if err != nil {
+			return err
+		}
+		for done := true; done; done, err = insertPlan.Next() {
```
I have a feeling this loop is more properly placed inside the Next method (where instead of looping we just return insertPlan.Next()) but I don't know how to fire off the cycle for the Next loop for createTableNode (it currently never gets called)

RaduBerinde commented:
Thanks for working on this! LGTM modulo some comments

---

Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/create.go, line 288 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy47UPPNiLmwBqxdIj:-KPy47UQzoJiUkDRm8_0:bboj61b) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L288)):*
> ```Go
> 	var selectClause *parser.SelectClause
> 	switch ps := n.Rows.Select.(type) {
> 	default:
> ```

[nit] default typically comes last

---

*[sql/create.go, line 491 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy5v0_rN8XX5ZfdAAu:-KPy5v0a9zTzPw4By23z:buh23ol) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L491)):*
> ```Go
> 			desiredTypesFromSelect[i] = col.Typ
> 		}
> 		insertPlan, err := n.p.newPlan(&parser.Insert{Table: &n.n.Table, Rows: n.n.Rows}, desiredTypesFromSelect, false)
> ```

long line, I would pull out the `&parser.Insert{}` expression.

---

*[sql/create.go, line 504 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPxuabjusDaAILIvOu_-r1-504:-KPy4w9AoSqZR0mwi4Td:b-8bg5cs) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L504)):*
<details><summary><i>Previously, WillHaack (Will Haack) wrote…</i></summary>
> I have a feeling this loop is more properly placed inside the Next method (where instead of looping we just return insertPlan.Next()) but I don't know how to fire off the cycle for the Next loop for createTableNode (it currently never gets called)
</details>
@knz 

---

*[sql/create.go, line 851 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy4VnApgY-k7mj2onu:-KPy4VnApgY-k7mj2onv:b-2e8jiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L851)):*
> ```Go
> 
> // MakeTableDescIfAs is the MakeTableDesc method for when we have a table that is created with the CREATE AS format.
> func MakeTableDescIfAs(p *parser.CreateTable, parentID sqlbase.ID, resultColumns []ResultColumn) (sqlbase.TableDescriptor, error) {
> ```

[nit] line too long

---

*[sql/create_test.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy4Zepa1hI453QTd7Z:-KPy4Zepa1hI453QTd7_:b-ifkhf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create_test.go#L155)):*
> ```Go
> 	s, sqlDB, _ := serverutils.StartServer(t, params)
> 	defer s.Stopper().Stop()
> 	if _, err := sqlDB.Exec(`CREATE DATABASE testDB`); err != nil {
> ```

We have a wrapper that does the Fatal calls automatically, here's an example: https://github.com/cockroachdb/cockroach/blob/54986344933bff1bed0a0991d03bbc0ee19c45a0/sql/distsql/server_test.go#L43

---

*[sql/parser/create.go, line 557 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy5MrxBtqY7sh6FPGY:-KPy5MrxBtqY7sh6FPGZ:b-xb4e5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/parser/create.go#L557)):*
> ```Go
> type CreateTable struct {
> 	IfNotExists bool
> 	IfAsExists  bool
> ```

This name is terrible :) Should just be `As`. 
`IfNotExists` has that name because it's related to the `IF NOT EXISTS` statement..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-:-KPy6-9tGlUtK_xhBhJI:bkpho3g)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[sql/create.go, line 277 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy6kLM0sy8-Po_Jb8M:-KPy6kLOUl7FSj4FOFZx:bj147kz) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L277)):*
> ```Go
> 			return nil, err
> 		}
> 		err := selectPlan.expandPlan()
> ```

Don't call expandPlan() in the constructor. This is invalid. The selectPlan's expandPlan() must be called in the createTableNode's own expandPlan().

---

*[sql/create.go, line 293 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy7-Yixy1ahauoQfCN:-KPy7-YkAXqmcHG60dvZ:bg0of41) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L293)):*
> ```Go
> 		selectClause = ps
> 	case *parser.ParenSelect:
> 		selectClause = ps.Select.Select.(*parser.SelectClause)
> ```

Use a recursion or a loop here. This is because you can have an arbitrarily deeply nested select ((((((((((((select ... ))))))))))).

---

*[sql/create.go, line 319 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy7GoDI96SxpK9jvrw:-KPy7GoDI96SxpK9jvrx:bs5ebx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L319)):*
> ```Go
> func (n *createTableNode) expandPlan() error {
> 	if n.n.IfAsExists && n.insertPlan != nil {
> 		return n.insertPlan.expandPlan()
> ```

The insertPlan never exists by the time createTableNode.expandPlan() is called so this test will always fail. However you should call the selectPlan's expandPlan() here as I suggested above.

---

*[sql/create.go, line 491 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy7VAnVV85vxijf4x2:-KPy7VApAx9q0SZN1Jhu:b-jmtyo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L491)):*
> ```Go
> 			desiredTypesFromSelect[i] = col.Typ
> 		}
> 		insertPlan, err := n.p.newPlan(&parser.Insert{Table: &n.n.Table, Rows: n.n.Rows}, desiredTypesFromSelect, false)
> ```

Use n.p.Insert(....) using the `Insert` constructor from insert.go instead.

---

*[sql/create.go, line 504 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPxuabjusDaAILIvOu_-r1-504:-KPy7hxr1C68E5x_QEhL:bgj1iyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L504)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> @knz 
</details>
What you are doing here is fine for a first approximation. If you want to do this the way you suggest, then you must change the result of the method StatementType() for the node you are using for "CREATE TABLE AS" to return the special value `Rows` (instead of `DDL` as it does currently). That will ensure Next() is called.

For this I would very highly recommend you define a new planNode other than createTableNode to do the entire work (ie. a new node createTableAsNode) that may require to duplicate some logic, but that's ok since it would make the rest of the code more readable). Otherwise, you must modify the current implementation of createTableNode.StatementType() in stmt.go to use a conditional: if AS is used, then the method must return `Rows`, otherwise it must return `DDL`.

---

*[sql/create.go, line 519 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy8ngA9Z0ZA_rAU41P:-KPy8ngC4Wr_2brbHaAX:b-20nzgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L519)):*
> ```Go
> func (n *createTableNode) Columns() []ResultColumn {
> 	if n.n.IfAsExists {
> 		return n.insertPlan.Columns()
> ```

That's incorrect. 1) Currently your implementation of "CREATE TABLE AS" cannot be used as the data source in another SELECT/INSERT so there shouldn't be any result columns then 2) if you wanted to make "CREATE TABLE AS" become a source, then you would need to use `n.selectPlan.Columns()` not the insertPlan's columns.

---

*[sql/create.go, line 526 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy91aD3J6-NuSqrddh:-KPy91aEiKrpaPn_9Wt5:b2nreh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L526)):*
> ```Go
> func (n *createTableNode) Ordering() orderingInfo {
> 	if n.n.IfAsExists {
> 		return n.insertPlan.Ordering()
> ```

Same as above.

---

*[sql/create.go, line 533 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy93xrlGWoshdL49N6:-KPy93xssmTJk87lwVP2:bdmpowb) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L533)):*
> ```Go
> func (n *createTableNode) Values() parser.DTuple {
> 	if n.n.IfAsExists {
> 		return n.insertPlan.Values()
> ```

This one is actually correct :)

---

*[sql/create_test.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy5wPFFX6jm_b-NzaF:-KPy5wPHXokmhyWWuj0t:b-h14773) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create_test.go#L150)):*
> ```Go
> }
> 
> func TestCreateTableAs(t *testing.T) {
> ```

This test would be best implemented as an input file for TestLogic in sql/testdata.

---

*[sql/parser/create.go, line 561 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy6BpLB4aWybbdkFBz:-KPy6BpMCd-fVLjxpCdz:b-bwsqzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/parser/create.go#L561)):*
> ```Go
> 	Interleave  *InterleaveDef
> 	Defs        TableDefs
> 	Rows        *Select
> ```

Just use one attribute in here called "AsSource" with a pointer of type `*Select` (then drop the IfAsExists attribute entirely). The pointer set to nil  would then mean there is no AS clause.

---

*[sql/parser/sql.y, line 1494 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy6WmKXd-WaHT1vA0l:-KPy6WmLHwlr4k8Su2Vi:b-i4kgr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/parser/sql.y#L1494)):*
> ```Yacc
>     $$.val = &CreateTable{Table: $3.normalizableTableName(), IfAsExists: true, IfNotExists: false, Interleave: nil, Defs: nil, Rows: $5.slct() }
>   }
> | CREATE TABLE IF NOT EXISTS any_name AS select_stmt
> ```

Add a test for these two in TestParse (parse_test.go)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[sql/create.go, line 504 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPxuabjusDaAILIvOu_-r1-504:-KPyA99D_keSvl1rBxG9:b-d6xivb) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L504)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> What you are doing here is fine for a first approximation. If you want to do this the way you suggest, then you must change the result of the method StatementType() for the node you are using for "CREATE TABLE AS" to return the special value `Rows` (instead of `DDL` as it does currently). That will ensure Next() is called.
> 
> For this I would very highly recommend you define a new planNode other than createTableNode to do the entire work (ie. a new node createTableAsNode) that may require to duplicate some logic, but that's ok since it would make the rest of the code more readable). Otherwise, you must modify the current implementation of createTableNode.StatementType() in stmt.go to use a conditional: if AS is used, then the method must return `Rows`, otherwise it must return `DDL`.
</details>
I've thought about this more and actually as long as you don't also support RETURNING, then what you have now is exactly what should be done. this is because a CREATE TABLE *is* really a DDL statement (unless there's also a RETURNING) so there is absolutely no reason to involve the executor's machinery to call Next() and count the number of result rows. It would also confuse SQL clients. So just don't change this. (Although you could still use a different planNode than createTable to clean up the code; it's just that this piece of the logic is OK.)

Please do however put a comment here to explain this. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[sql/create.go, line 533 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy93xrlGWoshdL49N6:-KPyDNFJ0hHZgWEG401U:b-tcb0wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L533)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> This one is actually correct :)
</details>
Scratch my previous comment on this. Values() must always be coherent with Columns(). See my comment above and the comment below about RETURNING. PRobably both methods should do like CREATE without regard to the AS clause (until we implement RETURNING here)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/create_test.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy5wPFFX6jm_b-NzaF:-KPyULWueK9_Do8ANoWy:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create_test.go#L150)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> This test would be best implemented as an input file for TestLogic in sql/testdata.
</details>
:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[sql/create.go, line 331 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPyTkgrT1-e2DOVPZQg:-KPyTkgrT1-e2DOVPZQh:bd2ihuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L331)):*
> ```Go
> 		desc, err = MakeTableDescIfAs(n.n, n.dbDesc.ID, n.selectPlan.Columns())
> 	} else {
> 		desc, err = MakeTableDesc(n.n, n.dbDesc.ID)
> ```

there's probably more of this method you want to skip -- the fk resolve below for instance, or basically anything else that looks at `n.n.Defs` ?

---

*[sql/create_test.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy5wPFFX6jm_b-NzaF:-KPyDHVGE0ayH1y8r8dZ:b-rqxvo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create_test.go#L150)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> :thumbsup: 
</details>
ditto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

WillHaack commented:
Review status: 0 of 5 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[sql/create.go, line 277 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy6kLM0sy8-Po_Jb8M:-KPyCdVMMsKRzlHSNDoE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L277)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Don't call expandPlan() in the constructor. This is invalid. The selectPlan's expandPlan() must be called in the createTableNode's own expandPlan().
</details>
Done.

---

*[sql/create.go, line 288 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy47UPPNiLmwBqxdIj:-KPy7zx52zvBwTRbWKS4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L288)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> [nit] default typically comes last
</details>
Done.

---

*[sql/create.go, line 293 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy7-Yixy1ahauoQfCN:-KPyGxQae4J735GtKBYx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L293)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Use a recursion or a loop here. This is because you can have an arbitrarily deeply nested select ((((((((((((select ... ))))))))))).
</details>
Done.

---

*[sql/create.go, line 319 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy7GoDI96SxpK9jvrw:-KPyCbH1771kswFySVmn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L319)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> The insertPlan never exists by the time createTableNode.expandPlan() is called so this test will always fail. However you should call the selectPlan's expandPlan() here as I suggested above.
</details>
Done.

---

*[sql/create.go, line 491 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy5v0_rN8XX5ZfdAAu:-KPyHMeAC3JNYO4OFTSM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L491)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> long line, I would pull out the `&parser.Insert{}` expression.
</details>
Done.

---

*[sql/create.go, line 491 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy7VAnVV85vxijf4x2:-KPyHN49hfI5dYdSNeaJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L491)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Use n.p.Insert(....) using the `Insert` constructor from insert.go instead.
</details>
Done.

---

*[sql/create.go, line 504 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPxuabjusDaAILIvOu_-r1-504:-KPyBS3870DdwJNSEsrF:b-g8om9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L504)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I've thought about this more and actually as long as you don't also support RETURNING, then what you have now is exactly what should be done. this is because a CREATE TABLE *is* really a DDL statement (unless there's also a RETURNING) so there is absolutely no reason to involve the executor's machinery to call Next() and count the number of result rows. It would also confuse SQL clients. So just don't change this. (Although you could still use a different planNode than createTable to clean up the code; it's just that this piece of the logic is OK.)
> 
> Please do however put a comment here to explain this. :)
</details>
Check to make sure my comment makes sense please :)

---

*[sql/create.go, line 519 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy8ngA9Z0ZA_rAU41P:-KPyHycR-rbmicktTrBh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L519)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> That's incorrect. 1) Currently your implementation of "CREATE TABLE AS" cannot be used as the data source in another SELECT/INSERT so there shouldn't be any result columns then 2) if you wanted to make "CREATE TABLE AS" become a source, then you would need to use `n.selectPlan.Columns()` not the insertPlan's columns.
</details>
Done.

---

*[sql/create.go, line 526 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy91aD3J6-NuSqrddh:-KPyaSJzTNsxVjAT7nV3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L526)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Same as above.
</details>
Done.

---

*[sql/create.go, line 533 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy93xrlGWoshdL49N6:-KPyI7RqxPSs1bbQ0xOc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L533)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Scratch my previous comment on this. Values() must always be coherent with Columns(). See my comment above and the comment below about RETURNING. PRobably both methods should do like CREATE without regard to the AS clause (until we implement RETURNING here)
</details>
Done.

---

*[sql/create.go, line 851 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy4VnApgY-k7mj2onu:-KPy8kMdyWMDAB9PPCPh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L851)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> [nit] line too long
</details>
Done.

---

*[sql/parser/create.go, line 557 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy5MrxBtqY7sh6FPGY:-KPy71zV3Jpui3w7nPC5:b4j09vg) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/parser/create.go#L557)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> This name is terrible :) Should just be `As`. 
> `IfNotExists` has that name because it's related to the `IF NOT EXISTS` statement..
</details>
derp, fixed

---

*[sql/parser/create.go, line 561 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy6BpLB4aWybbdkFBz:-KPyMuUFkn5SDVtpgwpu:b-ldu21e) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/parser/create.go#L561)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Just use one attribute in here called "AsSource" with a pointer of type `*Select` (then drop the IfAsExists attribute entirely). The pointer set to nil  would then mean there is no AS clause.
</details>
Added a method As() which returns AsSource != nil

---

*[sql/create_test.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy5wPFFX6jm_b-NzaF:-KPyaGQsBDvqc_YkMAtD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create_test.go#L150)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> ditto.
</details>
Done.

---

*[sql/create_test.go, line 155 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy4Zepa1hI453QTd7Z:-KPyI9AjRuTZbVKf0wHM:bq69oi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create_test.go#L155)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> We have a wrapper that does the Fatal calls automatically, here's an example: https://github.com/cockroachdb/cockroach/blob/54986344933bff1bed0a0991d03bbc0ee19c45a0/sql/distsql/server_test.go#L43
</details>
Done. But it looks like we're removing this test anyways as per Raphael's comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 6 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[sql/create.go, line 331 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPyTkgrT1-e2DOVPZQg:-KPycjP3gri6MnZ-NEES:b-87pea7) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/create.go#L331)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> there's probably more of this method you want to skip -- the fk resolve below for instance, or basically anything else that looks at `n.n.Defs` ?
</details>
The thing is all of this code already resolves to a no-op when the corresponding parts of the syntax node are empty (`for ...  range n.n.Defs` with an empty Defs is an empty loop). It doesn't seem like we can avoid any more computation work by adding extra conditionals here.

---

*[sql/create.go, line 508 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPybxahFPgerANpU11s:-KPybxaih8Y6EZAfehrA:b-7e8v9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/create.go#L508)):*
> ```Go
> 		}
> 		// This loop is done here instead of in the Next method
> 		// since CREATE TABLE is a DDL statement.
> ```

Append this to this comment: `... and Executor only runs Next() for statements with type "Rows"`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-:-KPyd4yFhqtNEsItxLT0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[sql/create.go, line 840 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPzdEguGOgpgJBFWF8-:-KPzdEguGOgpgJBFWF80:b-gdmfjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/create.go#L840)):*
> ```Go
> // MakeTableDescIfAs is the MakeTableDesc method for when we have a table
> // that is created with the CREATE AS format.
> func MakeTableDescIfAs(p *parser.CreateTable, parentID sqlbase.ID,
> ```

is this ever used outside `sql` package?

---

*[sql/create.go, line 841 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPzdX6GqhFBSP9eoKiw:-KPzdX6GqhFBSP9eoKix:b6kwt35) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/create.go#L841)):*
> ```Go
> // that is created with the CREATE AS format.
> func MakeTableDescIfAs(p *parser.CreateTable, parentID sqlbase.ID,
> 	resultColumns []ResultColumn) (sqlbase.TableDescriptor, error) {
> ```

see our style guide second function sigs, but tl;dr for this one is all the args, and only the args, on one line.

---

*[sql/create.go, line 1012 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPzbmC1XG6AMvNCCKeS:-KPzbmC1XG6AMvNCCKeT:b-2yzhsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/create.go#L1012)):*
> ```Go
> 
> func initTableDesc(parentID sqlbase.ID) sqlbase.TableDescriptor {
> 	desc := sqlbase.TableDescriptor{}
> ```

can you just `return sqlbase.TableDescriptor{ParentID: parentID...}` ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[sql/testdata/create_as, line 1 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KQ-p8RxHnGHm847aJqo:-KQ-p8RxHnGHm847aJqp:be64ih0) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/testdata/create_as#L1)):*
> ```
> statement ok
> ```

Could you add a test that ensures AS OF SYSTEM TIME fails with an error about AS OF SYSTEM TIME not being valid? Maybe:

```
statement error XXX
CREATE TABLE t AS SELECT * FROM stock AS OF SYSTEM TIME '2016-01-01'
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

WillHaack commented:
Review status: 2 of 6 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[sql/create.go, line 508 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPybxahFPgerANpU11s:-KQ0z6CjEMJfeCVIa3YG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/create.go#L508)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Append this to this comment: `... and Executor only runs Next() for statements with type "Rows"`.
</details>
Done.

---

*[sql/create.go, line 840 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPzdEguGOgpgJBFWF8-:-KQ0zTIoPVqjx7UKFjJz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/create.go#L840)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> is this ever used outside `sql` package?
</details>
Done.

---

*[sql/create.go, line 841 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPzdX6GqhFBSP9eoKiw:-KQ14h8AhHwWGxkodaF8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/create.go#L841)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> see our style guide second function sigs, but tl;dr for this one is all the args, and only the args, on one line.
</details>
Done.

---

*[sql/create.go, line 1012 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPzbmC1XG6AMvNCCKeS:-KQ15U5ksx0qDUUSOEgs:b-9nnf85) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/create.go#L1012)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> can you just `return sqlbase.TableDescriptor{ParentID: parentID...}` ?
</details>
Yeah, I also decided to just remove the method.

---

*[sql/parser/sql.y, line 1494 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KPy6WmKXd-WaHT1vA0l:-KQ1AgukE79Pev6sQzWi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/30ee249a84a675f0c7ab1f9f67e6fc60563695d1/sql/parser/sql.y#L1494)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Add a test for these two in TestParse (parse_test.go)
</details>
Done.

---

*[sql/testdata/create_as, line 1 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-KQ-p8RxHnGHm847aJqo:-KQ17oYu6X2tGU3A889C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/89ed20cc9513e98b8b19b8bee95f2aeb39367b8e/sql/testdata/create_as#L1)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Could you add a test that ensures AS OF SYSTEM TIME fails with an error about AS OF SYSTEM TIME not being valid? Maybe:
> 
> ```
> statement error XXX
> CREATE TABLE t AS SELECT * FROM stock AS OF SYSTEM TIME '2016-01-01'
> ```
> 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 6 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802)*
<!-- Sent from Reviewable.io -->

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8802#-:-KQ1OundcrbdcBCc8TP_:bnfp4nl)*
<!-- Sent from Reviewable.io -->
